Reduce do_3d_projection deprecation warnings in external artists

jakelishman opened the pull request:
## PR Summary

Previously, when finding the 3D zorder, external artists were assumed to
use the calling convention of `Patch3D.do_3d_projection` from Matplotlib
3.3, where it took a positional `renderer` parameter.  This triggered
a deprecation warning, even if the artist could support the new
convention.  Now we attempt to bind the signature first, and only emit a
deprecation warning if the calling convention does not appear to work.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

Fix #21740.

The original logic this modifies was introduced in #18302, where there was a comment that this was deliberately avoiding signature parsing to make things simpler.  However, with the new calling convention only being used if the `Artist.__module__` is `mplot3d`-internal, this meant that library code (e.g. Qiskit/qiskit-terra#7301) could not avoid triggering the deprecation warning, even if downstream `Artist` did actually fulfill the convention.  Since the warnings are triggered during the drawing phase, which may well be outside library control, we couldn't suppress the warnings without changing global user filters.

If the performance hit from using `signature.bind` is too heavy, it could also do
```python
try:
    return artist.do_3d_projection()
except TypeError as exc:
    # - check the exception stack that the issue was the bind, not raised within the function
    # - do the deprecation warning
    return artist.do_3d_projection(renderer)
```

tacaswell commented:
Thanks for working on this!  

This overlaps with https://github.com/matplotlib/matplotlib/pull/21704 where I took the more direct "try and see if it works" path.  @jakelishman Would you be willing to extract the tests from that PR to this one?

I think this PR has a marginally better implementation because it drops the special-case for our internal artists, but we still need to adjust the deprecation version.

jakelishman commented:
Ah sorry, I missed #21704 when I was searching to see if the issue was already reported.  I've incorporated the same sort of tests in 53e124b2ca, including the `match` argument to `pytest.warns` that @QuLogic had comment on that PR.  I added one for the split mpl 3.3/3.4 convention as well, since as a downstream library we're still looking to support 3.3 at the same time as 3.4 and 3.5.

tacaswell commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -412,24 +413,27 @@ def do_3d_projection(artist):
+                        "3.4",
+                        message="The 'renderer' parameter of "
+                        "do_3d_projection() was deprecated in Matplotlib "
+                        "%(since)s and will be removed %(removal)s.")
```
Can you also update the removal version?

jakelishman replied:
Sure - is it for removal in 3.6?

tacaswell replied:
Yes.  We should have removed it in 3.5, but that change is too big to do in a patch release so we have to delay it.

QuLogic replied:
Wait? 3.4 deprecation is scheduled for removal in 3.6. I think you've confused this with a different deprecation.

jakelishman replied:
As it stands, the message says it "was deprecated in Matplotlib 3.4 and will be removed two minor versions later". That's already consistent with 3.6, unless you'd rather it's more specific?

tacaswell replied:
Nevermind, @QuLogic is right I'm confused  🐑 .

QuLogic commented:
Thanks @jakelishman! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.
